### ISSUE TITLE
Disallow metadata on the type parameters of a function type?

### ISSUE THREAD

**Issue by eernstg** [dart-lang/language#656]
The common front end constant evaluator does not support metadata on a type parameter of a function type, and it turns out to be rather costly in terms of compilation time if this feature is implemented (because it implies that constant evaluation must traverse all types to find metadata _in_ them, not just _on_ them). Here is an example:

```dart
main() {
  void Function<@myMetadata X>(X x) myVariable = (_) {}; // Make this an error?
}
```

Note that this is concerned with function _types_ only, it doesn't affect type parameters that are declared by a generic function:

```dart
void myFunction<@myMetadata X>(X x) {} // OK.
```

A bit of searching in large bodies of source code doesn't produce any hits on likely forms where this feature is used. Also, it crashes the analyzer (cf. dart-lang/sdk# ).

So I'd propose that we consider making it a compile-time error and check that it doesn't occur, rather than rushing into an implementation of it in every tool.

@askeksa-google, you had some further comments on the performance implications?

WDYT, @leafpetersen, @munificent, @lrhn?

**Comment by askeksa-google** [comment]
An implementation of the constant transformation of these metadata can be seen here: https://dart-review.googlesource.com/c/sdk/+/123660

It is not a lot of code, but it makes the constant transformer visit many nodes that it didn't visit before.

**Comment by lrhn** [comment]
I don't think we should allow metadata anywhere except on declarations. Declarations have *identity*, so you can ask whether *that declaration* has metadata. 
The type variable of a function type, which is a *structural* type, doesn't seem like a declaration to me. It has no identity, because `X Function<X>(X)` and `Y Function<Y>(Y)` is considered the *same* type. Adding metadata to one version of the same type is meaningless.

I know that there is an opposite view which says that you should be able to put metadata on any syntactic construct - declarations, statements, expressions, so they can be found and processed by source processors, like it is an extra field on every node of the AST.
We should not do that unless we have a way to restrict annotations (like Java's retention policies), so our the VM won't have to keep everything until run-time.
(I guess dart2js can completely ignore metadata that it doesn't use itself since it doesn't have dart:mirrors).

**Comment by askeksa-google** [comment]
Is this change something that we would need to take through the breaking change process (given that none of our tools handle the situation properly to begin with)?

**Comment by askeksa-google** [comment]
Any decision on this? The CL to disallow these in the CFE is ready: https://dart-review.googlesource.com/c/sdk/+/123732

**Comment by eernstg** [comment]
Note that the following kind of metadata is equally expensive, but it doesn't crash the analyzer and there may be more pressure in the direction of allowing it:

```dart
const required = 1;

main() {
  Function({@required int i}) f = ({int i}) {};
}
```

@askeksa-google mentioned that we could choose to allow such annotations (at least the ones that currently don't crash any tools), but eliminate them before they are passed on to the backends. This would allow our tools to tolerate them, but ignore them, which might be more practical than making it a compile-time error.

Also, this would allow other tools (not based on the common front end and not using the analyzer) to use this kind of metadata for whatever purpose they may see fit.

**Comment by leafpetersen** [comment]
We've decided to disallow this.

**Comment by askeksa-google** [comment]
Somewhat related, the CFE currently ignores annotations on type parameters of functions that are not local: https://github.com/dart-lang/sdk/issues/28981

**Comment by eernstg** [comment]
Based on discussions with @askeksa-google, the performance implications that motivated the [decision](https://github.com/dart-lang/language/issues/656#issuecomment-553612389) to disallow metadata on the type parameters of a function type apply equally to _all_ expressions in a type.

In particular, it is costly because the constant evaluator needs to visit all types if each of them could contain metadata.

This means that we may also wish to disallow metadata

The analyzer does not implement the error for metadata on a type parameter of a function type. The error is reported by the CFE.

So reporting this error in the analyzer may be considered non-breaking because the CFE is likely to have eliminated such programs.

But adding an error for metadata also on normal parameters of a function type (that is, on formal parameters that are not type parameters) in the CFE as well as the analyzer would be breaking.

However, the behavior of the 'dart:mirrors' library seems to imply that metadata on normal parameters of a function type is ignored, which could be seen as an argument that this should be turned into an error as well.

**Comment by lrhn** [comment]
I don't think `dart:mirrors` is the primary consumer of metadata these days. It's more likely that the metadata is interpreted by source code reading tools like the analyzer or code generators.

That said, I don't think it makes sense to put metadata on things that do not have an identity.
A typedef declaration itself is fine, it has a name, but the function type it declares is structural, so it has no identity to distinguish it from a similar function type with different annotations. It makes no more sense than it does to annotate "the number 4".

**Comment by eernstg** [comment]
**Decided** at the language meeting Aug 12: The error should not be emitted, cf. https://github.com/dart-lang/sdk/issues/43039.
